### ISSUE TITLE
[Feature]: New API Datasource Method, datasource.update(property, value)

### ISSUE THREAD

**Issue by GreenFlux** [appsmithorg/appsmith#12405]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Request the ability to update an authenticated datasource by calling a method: 
`datasource1.update(property, value)`

We have access to individual API methods, like api1.run(), api1.clear(), etc, but we can't reference the parent datasource when an authenticated datasource is used. I would like to be able to update properties of the datasource via a function, so that updates can be triggered by user events, instead of only in the editor. 

### Why should this be worked on?

```
refreshTokenAPI.run()
newToken = refreshTokenAPI.data.access_token;
datasource1.update('headers.access_token',newToken)
```
This would allow updating the access token in an authenticated datasource-- at the datasource level, so that all related APIs inherit the correct token. 

`datasource1.update('url','baseurl/api/VERSION')`
This could be used to push a button in the app and switch between environments for A/B testing or connecting to different servers of the same service. 

`datasource1.update('headers.api_key', Input1.text)`
If I build an app that's intended to be copied and redeployed multiple times for different environments/clients, each one has to be opened in the app editor, updated and redeployed to change the datasource. This feature would allow the datasource to be updated by an end user, and in the live app, not the editor. Developers could build a modal and update the datasource with onSumbit, after asking the user for their API key.

**Comment by Nikhil-Nandagopal** [comment]
@GreenFlux I agree with the problem statement but I think the solution can be to enable dynamic bindings in the datasource or enabling multiple environments that can be switched on the fly

**Comment by GreenFlux** [comment]
Thanks @Nikhil-Nandagopal, I wasn't sure if dynamic bindings would be possible with authenticated datasources since the credentials are stored on the server and encrypted, instead of being sent to each client. If we use dynamic bindings, would one user be able to update the API key in a way that other users/clients could see and use the new key? Or would it only work for that one user?

**Comment by Nikhil-Nandagopal** [comment]
@GreenFlux that would be a problem with this approach as well since in this case too it would be sent from the client and could be a potential security problem. If it becomes dynamic, it could work either way where you define
- n environments of the datasource
- update the environment on the fly
- If you want to share the environment update then you store it in a database so that it updates for everyone
This way no secrets are being sent from the client and only the environment is being updated

**Comment by rohan-arthur** [comment]
reach 12000

**Comment by ajinkyakulkarni** [comment]
## Stats
| Stat | Values |
| ------------- | ------------- |
| Reach     | 4000 |
| Effort (months)  | 1 |

**Comment by ShawnFumo** [comment]
We've run into this issue as well. We're starting to use AppSmith for ops-related activities and need access to different environments and services, but don't want to use service accounts, since we want tracking of who is doing what. 

The most ideal situation for us would be a kind of datasource that authenticates on-the-fly for the current user, first going through the actual Authorization Code flow, and then using the refresh token specific to that user going forward until it expires. That way the user accesses the API with their own JWT and can also switch between different datasources to access different environments, only needing to auth when necessary.

Right now we are using the OIDC integration to log in a user for a particular environment and pass the token via <<>> syntax, but there is only one OIDC per instance, so now we need a different instance per environment. This isn't the end of the world for say Staging vs Production (though it'd still be nice to have the option to do it in one instance), but it'll break down as soon as we need to integrate with more things. Like say we want to access a file from a user's own OneDrive or get a list of their assigned stories from Jira. When they go to use that feature, it could auth against the service on the fly for them (using a popup window or iframe) and now they can access what they need.

I know it wouldn't be trivial to implement this, but it should be possible to do in a secure way. The OIDC login holds onto the token securely as part of the user session, so it should be possible to have more tokens in that same session. And if I'm remembering correctly, the current authenticated datasource auths without redirecting the whole page or restarting the app, so hopefully a similar process would work for on-the-fly auth, just that it isn't in the settings screen.
